Improve examples

galou opened the pull request:
#49823b9 is particularly important and should be cherry-picked, in the case the rest were to be rejected.

I read about the Braille example and wanted to give it a try. The result is not bad but the only thing that this script shows and that is not shown in other examples is the use of a custom class as input of the `pushPoints` method.

On the opposite, I think that the `numpy` example is pretty useful while being simple because it shows the flexibility of cadquery.

dcowden commented:
Thanks for contributing these! I love the numpy example! I'm seeing more
and more that CQ really fits will with the kind of problem space in which
numpy would likely already be in use, so that's great.

thanks for hacking on CQ !

On Mon, Oct 12, 2015 at 4:06 PM, Gaël Ecorchard <notifications@github.com>
wrote:

> #49823b9 is particularly important and should be cherry-picked, in the
> case the rest were to be rejected.
>
> I read about the Braille example and wanted to give it a try. The result
> is not bad but the only thing that this script shows and that is not shown
> in other examples is the use of a custom class as input of the pushPoints
> method.
>
> On the opposite, I think that the numpy example is pretty useful while
> being simple because it shows the flexibility of cadquery.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/jmwright/cadquery-freecad-module/pull/61
> Commit Summary
>
>    - Add an example with a polyline from a numpy array
>    - Remove flipLid parameter from Parametric Enclosure
>    - Support unicode in scripts
>    - Add Braille example
>
> File Changes
>
>    - *M* CadQuery/Examples/Ex023_Parametric_Enclosure.py
>    <https://github.com/jmwright/cadquery-freecad-module/pull/61/files#diff-0>
>    (5)
>    - *A* CadQuery/Examples/Ex028_Numpy.py
>    <https://github.com/jmwright/cadquery-freecad-module/pull/61/files#diff-1>
>    (26)
>    - *A* CadQuery/Examples/Ex029_Braille.py
>    <https://github.com/jmwright/cadquery-freecad-module/pull/61/files#diff-2>
>    (183)
>    - *M* CadQuery/Gui/Command.py
>    <https://github.com/jmwright/cadquery-freecad-module/pull/61/files#diff-3>
>    (3)
>
> Patch Links:
>
>    - https://github.com/jmwright/cadquery-freecad-module/pull/61.patch
>    - https://github.com/jmwright/cadquery-freecad-module/pull/61.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jmwright/cadquery-freecad-module/pull/61>.
>

jmwright commented:
I agree, these are great additions. I know some people who are going to love the numpy example. I've tinkered with the braille example before, but yours is much more complete and robust. Thanks!

jmwright commented:
@galou The braille example does cause some trouble on my system that's probably related to encoding. The code is declared to be UTF-8, and I get the following stack trace when I try to run it. I tried prefixing the string with the 'u' character, but it didn't help. We can move this out into a separate issue if there's not a quick fix.

```python
Running the Python command 'CadQueryExecuteScript' failed:
Traceback (most recent call last):
  File "/usr/lib/freecad/Mod/CadQuery/Gui/Command.py", line 178, in Activated
    imp.load_source('temp.module', tempFile.name)
  File "/tmp/tmp_x9fPZ", line 183, in <module>
    show(make_embossed_plate(text_lines, _cell_geometry))
  File "/tmp/tmp_x9fPZ", line 147, in make_embossed_plate
    base = make_base(text_lines, cell_geometry, base_thickness)
  File "/tmp/tmp_x9fPZ", line 133, in make_base
    base = cq.Workplane('XY').box(base_width, base_height, base_thickness,
  File "/usr/lib/freecad/Mod/CadQuery/Libs/cadquery/CQ.py", line 918, in __init__
    raise ValueError(" Provided value %s is not a valid work plane." % str(inPlane))

 Provided value XY is not a valid work plane.
```

galou commented:
Oups, sorry, for the development, I had deleted the embedded cadquery library and replaced in with a symlink to the cadquery directory from https://github.com/dcowden/cadquery. I just wrote a PR in it with the changes that make this work (hopefully).

galou commented:
I'm actually convinced that both repositories should be more clearly separated but that's another topic.

jmwright commented:
> I'm actually convinced that both repositories should be more clearly separated but that's another topic.

Agreed. I think that I'm just going to have to make CadQuery work as a submodule, even if it doesn't work how I think it should. I'll create an issue for that right now.

jmwright commented:
@galou I recently migrated the examples into the core cadquery library while converting them to the new method of running scripts that's portable between execution environments. There was an iditic points bug that was introduced somewhere along the way that had to be fixed in several examples. In the process of clean-up and migration the question came up as to how the numpy example you contributed should be working. The current script is this:

```python
import numpy as np
import cadquery as cq

# Square side and offset in x and y.
side = 10
offset = 5

# Define the locations that the polyline will be drawn to/thru.
# The polyline is defined as numpy.array so that operations like translation
# of all points are simplified.
pts = np.array([
    (side, 0),
    (side, side),
    (0, side),
    (0, 0),
]) + [offset, offset]

result = cq.Workplane('XY') \
    .polyline(pts).close().extrude(2) \
    .faces('+Z').workplane().circle(side / 2).extrude(1)

# Render the solid
show_object(result)
```
And the result given is this:

![screenshot_2017-10-13_08-50-42](https://user-images.githubusercontent.com/1015439/31546994-a4c5f2ac-aff3-11e7-80fb-6cee9cfaaa87.png)

Is the example still meeting the original intent, or has it gotten broken? If it's broken, please let me know what needs to be done to fix it.

galou commented:
Sorry, I just don't remember how it looked like but I guess it's not correct because we should have a square with a cylinder centered on top of it. It looks that the polyline is defined correctly.

Some definition of `pts` in the past was
```python
pts = np.array([
    (0, 0),
    (side, 0),
    (side, side),
    (0, side),
    (0, 0),
]) + [offset, offset]
```

It seems that you miss the first `(0, 0)`. It looks ok if you add it. Thanks!

jmwright commented:
When I add the (0,0) back in I get this result.

![screenshot_2017-10-15_21-24-35](https://user-images.githubusercontent.com/1015439/31591610-5313844e-b1ef-11e7-8fcf-579dddab66f3.png)

It seems like the polyline function is not respecting the offset that's being added to the numpy array. It's always trying to start from (0, 0). I could add a moveTo call and tack a 0 on for the z-axis, but it seems like there's a lower-level problem here that should be addressed. The polyline shouldn't have to start at 0, 0.

jmwright commented:
I think the problem is [right here](https://github.com/dcowden/cadquery/blob/484a1ca667b79a773c7ecc17a17b6fabb437cc06/cadquery/cq.py#L1695) in the CadQuery library. I need to figure out if it's possible to safely change this to be the first point of the polyline point array rather than the last point from the last object in the stack.

galou commented:
This is strange, I get the correct result without problem.

![screenshot from 2017-10-18 13 55 10](https://user-images.githubusercontent.com/2596352/31717490-612acd02-b40c-11e7-908f-5b36adfb53c2.png)

```python
import numpy as np
import cadquery as cq

# Square side and offset in x and y.
side = 10
offset = 5

# Define the locations that the polyline will be drawn to/thru.
# The polyline is defined as numpy.array so that operations like translation
# of all points are simplified.
pts = np.array([
    (0, 0),
    (side, 0),
    (side, side),
    (0, side),
    (0, 0),
]) + [offset, offset]

result = cq.Workplane('XY') \
    .polyline(pts).close().extrude(2) \
    .faces('+Z').workplane().circle(side / 2).extrude(1)

# Render the solid
show_object(result)
```
Commit 0a6d2e9a31df6ce9704c39875e452ba102e71100

OS: Ubuntu 14.04.5 LTS
Word size of OS: 64-bit
Word size of FreeCAD: 64-bit
Version: 0.17.12412 (Git)
Build type: None
Branch: master
Hash: 4a8cd391b53b8d6d04026bf1f85c32002989e06e
Python version: 2.7.6
Qt version: 4.8.6
Coin version: 4.0.0a
OCC version: 7.1.0
Locale: English/UnitedStates (en_US)

jmwright commented:
Interesting, it works fine in 0.17 for me too. I was trying it in 0.16 before. I spent about 30 minutes last night trying to make the example work in 0.16 without changing the intent of the example, or doing something that was a hack. I didn't have much luck. I'm tempted to revert the example back to what you have posted above and put a note at the top that warns users that it may not run correctly in 0.16.

jmwright commented:
I ended up doing what I was suggesting in the last post.

https://github.com/dcowden/cadquery/blob/master/examples/FreeCAD/Ex028_Numpy.py#L1

galou commented:
Good choice. I personally consider FreeCAD v0.16 obsolete even if it's less buggy than v0.17 (for now).
